### PR TITLE
commit of vsphere provider changes

### PR DESCRIPTION
**Pull request by cooktheryan**
Replaces #11909 @dav1x and @vikram-redhat

### CODE REVIEW

**Comment by dav1x** [comment]
LGTM!

**Comment by vikram-redhat** [comment]
@xltian - can you assign someone to do the QE review?

**Comment by vikram-redhat** [comment]
@ahardin-rh - can you take a look once we have QE approval?

**Comment by jhou1** [comment]
This looks great!

File: install_config/topics/configuring_for_vsphere.adoc [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+
+The `openshift_cloudprovider_vsphere_` values are required for {product-title}
+to be able to create `vSphere` resources such as VMDKs on datastores for
+persistent volumes.
+
+[source,yaml]
+----
+$ cat /etc/ansible/hosts
+
+[OSEv3:children
```
Comment (liangxia) [inline install_config/topics/configuring_for_vsphere.adoc]: Missing `]` at end of line

File: install_config/topics/configuring_for_vsphere.adoc [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+openshift_deployment_type=openshift-enterprise
+# Required per https://access.redhat.com/solutions/3480921
+oreg_url=registry.access.redhat.com/openshift3/ose-${component}:${version}
+openshift_examples_modify_imagestreams=true
+openshift_storage_glusterfs_image=registry.access.redhat.com/rhgs3/rhgs-server-rhel7
+openshift_storage_glusterfs_block_image=registry.access.redhat.com/rhgs3/rhgs-gluster-block-prov-rhel7
+openshift_storage_glusterfs_s3_image=registry.access.redhat.com/rhgs3/rhgs-s3-server-rhel7
+openshift_storage_glusterfs_heketi_image=registry.access.redhat.com/rhgs3/rhgs-volmanager-rhel7
+
+VM deployment
```
Comment (liangxia) [inline install_config/topics/configuring_for_vsphere.adoc]: This should be a comment, which needs start with `#`

File: install_config/topics/configuring_for_vsphere.adoc [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+
+default_subdomain=example.com
+openshift_master_cluster_hostname=openshift.example.com
+openshift_master_cluster_public_hostname=openshift.example.com
+openshift_master_default_subdomain=apps.example.com
+
+os_sdn_network_plugin_name='redhat/openshift-ovs-networkpolicy'
+osm_use_cockpit=true
+
+# red hat subscription name and password
```
Comment (liangxia) [inline install_config/topics/configuring_for_vsphere.adoc]: I was told is should be always `Red Hat`

File: install_config/topics/configuring_for_vsphere_registry.adoc [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+----
+VMFS
+cd /vmfs/volumes/datastore1/
+# VSAN
+cd /vmfs/volumes/vsanDatastore/
+
+# VMFS
+mkdir kubevols # Not needed but good hygiene
+# VSAN
+/usr/lib/vmware/osfs/bin/osfs-mkdir kubevols # Needed
```
Comment (liangxia) [inline install_config/topics/configuring_for_vsphere_registry.adoc]: I think it is cleaner to write as below.

```
# VMFS
cd /vmfs/volumes/datastore1/
mkdir kubevols # Not needed but good hygiene

# VSAN
cd /vmfs/volumes/vsanDatastore/
/usr/lib/vmware/osfs/bin/osfs-mkdir kubevols # Needed
```

File: install_config/topics/configuring_for_vsphere_storage.adoc [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+== Provisioning VMware vSphere volumes via a Storage Class
+
+. {product-title} creates the following `storageclass` when you use the `vsphere-volume`
+provisioner and if you use the `openshift_cloudprovider_kind=vsphere` and
+`openshift_vsphere_*` variables in the Ansible inventory. Otherwise, you can create
+it manually:
++
+[source,yaml]
+----
+$ oc get --export storageclass standard -o yaml
```
Comment (liangxia) [inline install_config/topics/configuring_for_vsphere_storage.adoc]: The name `standard` here should be replaced with `vsphere-standard` to be consistent.

File: install_config/topics/manually_configuring_for_vsphere.adoc [hunk 6]
```diff
@@ -0,0 +1,190 @@
...
++
+[source,yaml]
+----
+kubernetesMasterConfig:
+  ...
+  apiServerArguments:
+    cloud-provider:
+      - "vsphere"
+    cloud-config:
+      - "/etc/origin/cloudprovider/vsphere.conf"*
```
Comment (liangxia) [inline install_config/topics/manually_configuring_for_vsphere.adoc]: Extra `*` should be removed.

File: install_config/topics/manually_configuring_for_vsphere.adoc [hunk 7]
```diff
@@ -0,0 +1,190 @@
...
+  controllerArguments:
+    cloud-provider:
+      - "vsphere"
+    cloud-config:
+      - "/etc/origin/cloudprovider/vsphere.conf"*
```
Comment (liangxia) [inline install_config/topics/manually_configuring_for_vsphere.adoc]: Same here.

File: install_config/topics/manually_configuring_for_vsphere.adoc [hunk 8]
```diff
@@ -0,0 +1,190 @@
...
+configuration map] and update the contents of the `*kubeletArguments*`
+section:
++
+[source,yaml]
+----
+kubeletArguments:
+  cloud-provider:
+    - "vsphere"
+  cloud-config:
+    - "/etc/origin/cloudprovider/vsphere.conf"*
```
Comment (liangxia) [inline install_config/topics/manually_configuring_for_vsphere.adoc]: Same here.

**Comment by cooktheryan** [comment]
thanks @liangxia ! I just pushed some updates

**Comment by liangxia** [comment]
LGTM

**Comment by cooktheryan** [comment]
@vikram-redhat  both

**Comment by vikram-redhat** [comment]
@cooktheryan thanks. Looks like there are some build errors due. Can you take a look and update?

https://travis-ci.org/openshift/openshift-docs/builds/430680652#L526 and https://travis-ci.org/openshift/openshift-docs/builds/430680652#L627

File: install_config/topics/configuring_for_vsphere.adoc [hunk 9]
```diff
@@ -0,0 +1,174 @@
...
+----
+[OSEv3:vars]
+openshift_cloudprovider_kind=vsphere
+openshift_cloudprovider_vsphere_username=administrator@vsphere.local <1>
+openshift_cloudprovider_vsphere_password=<password>
+openshift_cloudprovider_vsphere_host=10.x.y.32 <2>
+openshift_cloudprovider_vsphere_datacenter=<Datacenter> <3>
+openshift_cloudprovider_vsphere_datastore=<Datastore> <4>
+----
+<1> The username with the appropriate permissions to create and attach disks in vSphere.
```
Comment (ahardin-rh) [inline install_config/topics/configuring_for_vsphere.adoc]: user name

File: install_config/topics/configuring_for_vsphere.adoc [hunk 10]
```diff
@@ -0,0 +1,174 @@
...
+<2> The vCenter server address.
+<3> The vCenter Datacenter name where the {product-title} VMs are located.
+<4> The datastore that will be used for creating VMDKs.
```
Comment (ahardin-rh) [inline install_config/topics/configuring_for_vsphere.adoc]: The datastore used for creating VMDKs

File: install_config/topics/configuring_for_vsphere_registry.adoc [hunk 11]
```diff
@@ -0,0 +1,98 @@
+// Module included in the following assemblies:
+//
+// * install_config/configuring_for_vsphere.adoc
+
+[id='vsphere-configuring-registry_{context}']
+= Configuring the {product-title} registry for vSphere
+The steps below define the steps define the manual process to
```
Comment (ahardin-rh) [inline install_config/topics/configuring_for_vsphere_registry.adoc]: The following steps define the manual process of storage creation, which is used . . .

File: install_config/topics/configuring_for_vsphere_registry.adoc [hunk 12]
```diff
@@ -0,0 +1,98 @@
...
+openshift_hosted_registry_storage_access_modes=['ReadWriteOnce'] <2>
+openshift_hosted_registry_storage_annotations=['volume.beta.kubernetes.io/storage-provisioner: kubernetes.io/vsphere-volume'] <3>
+openshift_hosted_registry_replicas=1 <4>
+----
+<1> The storage type.
+<2> vSphere volumes only support `RWO`.
+<4> The annotation for the volume.
+<4> The number of replicas to configure.
+
+NOTE: The brackets in the configuration file above are required.
```
Comment (ahardin-rh) [inline install_config/topics/configuring_for_vsphere_registry.adoc]: Note boxes are formatted as:

    [NOTE]
    =====
    The brackets in the configuration file above are required.
    ====

See https://github.com/openshift/openshift-docs/blob/master/contributing_to_docs/doc_guidelines.adoc#admonitions

File: install_config/topics/configuring_for_vsphere_storage.adoc [hunk 13]
```diff
@@ -0,0 +1,191 @@
...
+====
+
+
+[discrete]
+== Prerequisites
+
+Before creating PVs using vSphere, ensure your
+{product-title} cluster meets the following requirements:
+
+* {product-title} must first be for vSphere Cloud Provider
```
Comment (ahardin-rh) [inline install_config/topics/configuring_for_vsphere_storage.adoc]: What does the first requirement mean?

File: install_config/topics/configuring_for_vsphere_storage.adoc [hunk 14]
```diff
@@ -0,0 +1,191 @@
...
+* Each node host in the infrastructure must match the vSphere VM name.
+* Each node host must be in the same resource group.
+
+[IMPORTANT]
+====
```
Comment (ahardin-rh) [inline install_config/topics/configuring_for_vsphere_storage.adoc]: I would take this procedure out of the IMPORTANT box. If this is a prerequisite procedure, we can list it as such.

File: install_config/topics/manually_configuring_for_vsphere.adoc [hunk 15]
```diff
@@ -0,0 +1,190 @@
...
+[id='vsphere-configuring-masters-manual_{context}']
+= Option 2: Manually configuring {product-title} for vSphere
+
+== Manually configuring master hosts for vSphere
+
+Perform the following on all master hosts.
+
+.Procedure
+
+. Edit the master configuration file at `/etc/origin/master/master-config.yaml`
```
Comment (ahardin-rh) [inline install_config/topics/manually_configuring_for_vsphere.adoc]: *_/etc/origin/master/master-config.yaml_*

File: install_config/topics/manually_configuring_for_vsphere.adoc [hunk 16]
```diff
@@ -0,0 +1,190 @@
...
+        scsicontrollertype = pvscsi <14>
+
+[Network]
+        public-network = "VM Network" <15>
+----
+<1> Any properties set in the `[Global]` section are used for all specified vcenters unless overriden by the settings in the individual `[VirtualCenter]` sections.
+<2> vCenter username for the vSphere cloud provider.
+<3> vCenter password for the specified user.
+<4> Optional. Port number for the vCenter server. Defaults to port `443`.
+<5>  Set to `1` if the vCenter uses a self-signed cert.
```
Comment (ahardin-rh) [inline install_config/topics/manually_configuring_for_vsphere.adoc]: s/cert/certificate

File: install_config/topics/manually_configuring_for_vsphere.adoc [hunk 17]
```diff
@@ -0,0 +1,190 @@
...
+If an {product-title} node host goes down in zone "A" and the pods
+should be moved to zone "B".
+See https://kubernetes.io/docs/admin/multiple-zones/#limitations[Multiple zone
+limitations] in the Kubernetes documentation for more information.
+
+[IMPORTANT]
+====
+This ensures that the VMDK always presents a consistent UUID to the VM, allowing the disk
+to be mounted properly.
+
```
Comment (ahardin-rh) [inline install_config/topics/manually_configuring_for_vsphere.adoc]: Extra line

File: install_config/topics/manually_configuring_for_vsphere.adoc [hunk 18]
```diff
@@ -0,0 +1,190 @@
...
+====
++
+For every virtual machine node that will be participating in the cluster:
+VM properties -> VM Options -> Advanced -> Configuration Parameters -> disk.enabledUUID=TRUE
++
+Alternatively, link:https://github.com/vmware/govmomi/tree/master/govc[GOVC tool] can be used:
```
Comment (ahardin-rh) [inline install_config/topics/manually_configuring_for_vsphere.adoc]: Alternatively, the

File: install_config/topics/manually_configuring_for_vsphere.adoc [hunk 19]
```diff
@@ -0,0 +1,190 @@
...
+[source,bash]
+----
+govc vm.change -e="disk.enableUUID=1" -vm='VM Path'
+----
++
+[NOTE]
+====
+If {product-title} node VMs are created from a template VM, then
+`disk.EnableUUID=1` can be set on the template VM. VMs cloned from this
+template, inherit this property.
```
Comment (ahardin-rh) [inline install_config/topics/manually_configuring_for_vsphere.adoc]: template inherit this property

File: install_config/topics/vsphere_prerequisites.adoc [hunk 20]
```diff
@@ -0,0 +1,130 @@
...
+----
+
+.. Find the Node VM paths:
++
+[source,bash]
+----
+govc ls /datacenter/vm/<vm-folder-name>
+----
+
+.. Set disk.EnableUUID to true for all VMs:
```
Comment (ahardin-rh) [inline install_config/topics/vsphere_prerequisites.adoc]: Set `disk.EnableUUID` to `true` for all VMs:

**Comment by ahardin-rh** [comment]
@cooktheryan Just some minor suggestions from me. Once those are addressed and the build failures fixed, you can squash your commits down to one so that we can cleanly cherry-pick the changes to the correct product versions. Thanks for this!

File: install_config/topics/configuring_for_vsphere_storage.adoc (hunk 14 above)
Comment (cooktheryan) [reply install_config/topics/configuring_for_vsphere_storage.adoc]: @dav1x  can you provide information here if these steps are actually required or a way to meet the requirements from @ahardin-rh

File: install_config/topics/configuring_for_vsphere_storage.adoc (hunk 13 above)
Comment (cooktheryan) [reply install_config/topics/configuring_for_vsphere_storage.adoc]: @dav1x do you have a better rewrite for the above?

**Comment by ahardin-rh** [comment]
@cooktheryan I'm fetching your branch from your fork to troubleshoot. Please stand by  :timer_clock: :smile:

**Comment by ahardin-rh** [comment]
@cooktheryan Looks like when I run `asciibinder build` per [our tools and setup guidance](https://github.com/openshift/openshift-docs/blob/master/contributing_to_docs/tools_and_setup.adoc#install-asciibinder-and-dependencies), the note in Configuring the OpenShift Container Platform registry for vSphere is rendering as a heading and the  Manually configuring OpenShift Container Platform registry for vSphere section is not properly rendering, so I think the issue is with that somehow. Also, I am not able to see the vsphere_prerequisites, configuring_for_vsphere_registry, configuring_for_vsphere_storage, or other new topics in order to edit them. Sorry for all the confusion, but we're moving into a modularized docs structure. I think we need to move your new topics into the Modules folder (not a Topics folder). That may be messing with the build as well.

File: install_config/topics/configuring_for_vsphere_storage.adoc (hunk 13 above)
Comment (dav1x) [reply install_config/topics/configuring_for_vsphere_storage.adoc]: * {product-title} must first be configured for vSphere Cloud Provider (VCP)

File: install_config/topics/configuring_for_vsphere_storage.adoc (hunk 14 above)
Comment (dav1x) [reply install_config/topics/configuring_for_vsphere_storage.adoc]: Actually, can we remove this line completely? It's a best practice but not required.

**Comment by cooktheryan** [comment]
@ahardin-rh there was a small issue in the registry file that caused us issues. That has now been resolved.

**Comment by ahardin-rh** [comment]
@cooktheryan I took your branch and carried the work over into https://github.com/openshift/openshift-docs/pull/12306

I had to fix some formatting and squash down the commits there, so I will close this PR in favor of that one and get things moving along for publishing. Thanks for all of your hard work on this!

cc @vikram-redhat

**Pull request closed without merging**